3.1 - Shell "requested" param

HavokInspiration opened the pull request:
Dispatching a Shell within another shell (with `Shell::dispatchShell()`) makes the welcome message called twice :

```
Welcome to CakePHP v3.0.0 Console
---------------------------------------------------------------
App : src
Path: /home/yves/PhpStormProjects/cakephp-app/src/
---------------------------------------------------------------

First Shell output
Need to dispatch another shell inside this one

Welcome to CakePHP v3.0.0 Console
---------------------------------------------------------------
App : src
Path: /home/yves/PhpStormProjects/cakephp-app/src/
---------------------------------------------------------------

Second shell output
```

This PR adds a new parameter called "requested" that will prevent the welcome message to be displayed.
By extension, this parameter could be used in userland to detect if the Shell was called with the dispatchMethod.
Additionally, any call to dispatchShell() will append this parameter to the list of parameters.

More details on #6272.

HavokInspiration commented:
I'll fix the failing tests

markstory commented on src/Console/Shell.php:
```diff
@@ -211,7 +211,10 @@ public function initialize()
     {
-        $this->_welcome();
+        $requested = $this->param('requested');
+        if (empty($requested)) {
```
Couldn't you just test the truthiness of the param? Given that its always locked to a boolean it will be a yes/no condition.

HavokInspiration replied:
Makes sense.

markstory commented on src/Console/Shell.php:
```diff
@@ -307,6 +310,7 @@ public function dispatchShell()
         if (is_string($args[0]) && count($args) === 1) {
             $args = explode(' ', $args[0]);
         }
+        $args[] = '--requested';
```
I wonder if we could skip adding a new cli flag to all shells if instead we supported 'extra' parameters like requestAction() does. It might get a bit twisty though as ShellDispatcher and Shell would have to know about these new extra parameters.

HavokInspiration replied:
Something more tied to the root ``Shell`` class ?
Like the ``bare`` parameter used to prevent layouts from being rendered when a request is dispatched with ``requestAction()`` ?

I can take a look in that direction and see what I can come up with.

HavokInspiration replied:
@markstory Took a deeper look at your suggestion. Something like that could work :

- Change dispatchShell() so it can accept an array of parameters allowing those extra parameters to be passed :
```php
Shell::dispatchShell([
    'command' => 'schema create DbAcl'
    'extra' => [
        'requested' => true
    ]
])
```
``command`` would be parsed as the current arguments are parsed and ``extra`` passed to the ``ShellDispatcher`` (which would have its constructor modified to accept them as well as a new property to store them until the ``Shell`` is dispatched)
Current signature could still work.
- Extra parameters would be passed to the Shell either through its constructor or a dedicated method called before dispatch.
- Have a new ``extra`` property on ``Shell`` (or a new dimension in ``param``)
- Make ``Shell::param()`` able to access ``extra`` param via something like ``Shell::param('requested', 'extra')``. 

This way, no new flags are added and it is possible to add others in the future and gives the ability to users to have their own extra parameters for their Shells without having to create a new flag.
This would however allow those extra parameters to be used only with ``dispatchShell()`` and prevent users from using them in a "normal call" (for instance, with the current states of the changes, it is possible to call ``cake my_shell my_command --requested`` and not have the welcome message displayed). But I don't know if it has any real value though so I'm okay with that.

This has not been tested or coded. Just a draft of ideas. What do you think ?

markstory replied:
The extra parameters could be merged into the arguments/params after the options are parsed. That would make the code inside a shell not need to know how it got its parameters.

If the parameters are not present in the option parser I don't think one could use them from the cli interface.

HavokInspiration committed:
Allow Shells to be dispatched with a new "extra" arguments

This "extra" arguments allow additional parameters to be passed to the Shell without them needing to be a CLI flag or a command argument.
 This commit also introduce a "requested" extra parameter that is automatically passed when a Shell is dispatched with a Shell::dispatchShell() call. This parameter, when set and not null or false, prevents the default cake shell welcome message from being displayed

HavokInspiration commented:
@markstory I've refactored it all and went with your suggestion of having a new ``extra`` arguments that can be used when dispatching a Shell. I went with passing the parameter directly to ``runCommand()`` and ``dispatch()``. Felt simpler and avoid having the extra parameters stored somewhere. And since they are "runtime" parameter in a way, that makes more sense IMO.

It might be good to test the parsing of the ``dispatchShell()`` method. But I'll have to admit that I need some hint on how to do this, since all it returns is a ``bool``.

markstory commented on src/Console/Shell.php:
```diff
@@ -211,7 +211,9 @@ public function initialize()
     public function startup()
     {
-        $this->_welcome();
+        if (!(bool)$this->param('requested')) {
```
`!` does implicit boolean casting for you.

markstory commented:
Testing `dispatchShell()` is pretty painful which is why there aren't any existing tests sadly.

HavokInspiration commented:
Ok, that's what I was afraid of.
What about having a method to do [this part of the logic](https://github.com/HavokInspiration/cakephp/blob/5db6dc3ff691949ad6ace306eb4a0162d0c3934f/src/Console/Shell.php#L310-L318) and add tests for it ? That would at least prevent regressions in arguments parsing if ``dispatchShell()`` needs modifications in the future.

markstory commented:
@HavokInspiration Another method sounds like a good option to me.

HavokInspiration committed:
Add a method that will be in charge of parsing arguments passed to a Shell::dispatchShell() call.
Add tests for that method.

HavokInspiration commented:
Added a method to parse arguments given to ``dispatchShell()`` and corresponding tests. 
I also did a "real" test dispatching a shell from another shell and it works okay.

It does feel flying without wings to not have tests validating that ``dispatchShell()`` works okay. I certainly would not want to be the guy who broke a part of the framework :smile: 

Is there no way to test this? With I don't know... A custom test shell that will overwrite ``out()`` to make it log the output, call a command that dispatch the same shell that calls ``out()`` and compares the final values of the outputed message and what we expect ? 
Something like this :

```php
class TestingShell extends Shell
{
    
    public $out = [];
    
    public function out($message = null, $newlines = 1, $level = Shell::NORMAL)
    {
        $this->out[] = $message;
    }

    public function testTask()
    {
        $this->out('I am a test task, I dispatch another Shell');
        $this->dispatchShell('testing dispatch_test_task');
    }

    public function dispatchTestTask()
    {
        $this->out('I am a dispatched Shell');
    }
}

//In test files
$TestingShell = new TestingShell();
$TestingShell->runCommand('test_task');

//Compare $TestingShell->out with a built array
```

It's a draft done quick and dirty, just to have something more concrete than my so poorly written explanation. But that might give some sort of coverage to ``dispatchShell()``. Or maybe I'm missing something and that would not work at all.

markstory commented:
Looks good to me.

HavokInspiration commented:
@markstory I think this was merged too early. Since master was merged to 3.1, the dispatchShell test added on #6325 will probably fail (the welcome message will not be outputed, hence the expected value will not be correct anymore.
I'll try to submit a quick fix for the test for the 3.1 branch with just github online editor.
I'll add more in depth tests to dispatchShell linked to this PR (hopefully, before the end of the week)

markstory commented:
Whoops. Failing tests can be fixed easily enough :smile:
